### ISSUE TITLE
VDIF time encoding fails to account for leap seconds

### ISSUE THREAD

**Issue by andrerenard** [kotekan/kotekan#498]
It appears that the implementation of the VDIF time code in both the ARO implementation `iceBoardVDIF.hpp` and the CHIME `pulsarPostProcess.cpp` fail to correctly account for leap seconds. 

This is because they both use a unix time stamp either from the system unix time or the GPS unix time from `fpga_master` to set the initial offset.  This works as long as there are no leap seconds between the chosen VDIF epoch (`ref_epoch`) and either the current system time or the frame0 GPS/FPGA time.   However since unix time “loses” leap seconds by fixing the length of each day to 84600 seconds, this means a leap second after the reference epoch won’t get encoded in the initial offset.   This is a violation of the VDIF spec which requires that the VDIF `seconds` field encode all seconds including leap seconds since the reference epoch (which is defined in UTC).

Interestingly because the current code uses the FPGA as a relative clock after setting the initial offset to the reference epoch, any leap seconds that happen _while_ the system is running will be correctly encoded - until the system restarts and then loses knowledge of the leap second. 

The correct approach to this would be to convert the initial unix time (system or GPS) into TAI and then subtract the TAI version of the UTC reference epoch.  Since the unix<->TAI function changes as leap seconds are announced the best option seems to be reference the NTP leap seconds tables, which will get dynamically updated, instead of relying on keeping a TAI library with hardcoded leap seconds up to date.   That should correct the initial offset to include the (currently missing) leap seconds since the reference epoch.  And as noted above once the initial offset is correctly set the system will handle leap seconds which happen while running correctly. 

It’s worth noting: in the CHIME case since we chose an epoch in 2018, and there have been no leap seconds since, the CHIME VDIF headers are currently correct, but will fail on the first restart after the next leap second.   In the case of ARO and other single dish VDIF, there is currently a 5 second error resulting from the failure to add the leap seconds since the 2000 UTC epoch.

**Comment by jrs65** [comment]
Make sure you keep myself abreast of what changes you are doing here, as we need to have the correct time encodings for the cosmology end too. I need to think about what you're proposing to ensure it's compatible.

One thing to note is that one of the things that all observatories do (and we plan to do) is *never* run through a leap second. Partly because of issues in the real time side, but also because of problems accounting for it properly down stream. So I think we shouldn't plan for that eventuality.

**Comment by jrs65** [comment]
Also, note that the announcement has come out that there won't a leap second at the end of this year so there's at least another 9 months to fix it.

**Comment by andrerenard** [comment]
@jrs65 yes, I would actually like to talk to you about this at the hackathon this week. The change I'm making here doesn't impact the cosmology end (which I think is correct as long as you never run _through_ a leap second like you say).  However there is a way to run through a leap second by using some form of TAI, so we might be able to actually make the cosmology robust to leap seconds if we were willing to make changes to the cosmology pipeline both at the recording and analysis pipelines.    I also need to check with JF about what the GPS will do on a leap second day, if it slows the 10 MHz clock to force the number of seconds to 84600 seconds then we could be in for other strange effects.  Although my guess is the 10 MHz clock is locked to TAI.

**Comment by andrerenard** [comment]
This is a good reference as to what Unix/NTP/Chrony to do deal with leap seconds: https://developers.redhat.com/blog/2015/06/01/five-different-ways-handle-leap-seconds-ntp/

Annoyingly it looks like C++20 implements a TAI time lib, but C++14 doesn't have one in the STL: https://en.cppreference.com/w/cpp/chrono/tai_clock

**Comment by andrerenard** [comment]
Currently I'm thinking of using this: https://github.com/HowardHinnant/date to do time conversion.  It appears to correctly implement dynamic updates to the timezone files with leap second data.

**Comment by andrerenard** [comment]
That library includes a good description of this problem here: https://github.com/HowardHinnant/date/wiki/Examples-and-Recipes#unix_gps_epoch_diff
